File: plldb-server/src/main/java/cn/twincest/plldb/util/JSONUtil.java
package cn.twincest.plldb.util;

import com.google.gson.*;
import org.bson.types.ObjectId;

import java.lang.reflect.Type;

public class JSONUtil {

	private static class ObjectIdTypeAdapter implements JsonSerializer<ObjectId>, JsonDeserializer<ObjectId> {
		@Override
		public JsonElement serialize(ObjectId objectId, Type type, JsonSerializationContext jsonSerializationContext) {
			return new JsonPrimitive(objectId.toHexString());
		}

		@Override
		public ObjectId deserialize(JsonElement jsonElement, Type type, JsonDeserializationContext jsonDeserializationContext) throws JsonParseException {
			return new ObjectId(jsonElement.getAsString());
		}
	}

	private static final Gson GSON = new GsonBuilder()
			.disableHtmlEscaping()
			.serializeNulls()
			.setDateFormat("yyyy-MM-dd HH:mm:ss")
			.registerTypeAdapter(ObjectId.class, new ObjectIdTypeAdapter())
			.create();

	public static JsonElement toTree(Object o) {
		return GSON.toJsonTree(o);
	}

	public static <T> T fromTree(JsonElement e, Type type) {
		return GSON.fromJson(e, type);
	}

	public static <T> T fromString(String s, Type type) {
		return GSON.fromJson(s, type);
	}

}

File: plldb-server/src-refactor/main/java/cn/twin_kles/plldb/util/JSONUtil.kt
package cn.twin_kles.plldb.util

import org.bson.types.ObjectId
import cn.twin_kles.plldb.util.JSONUtil.ObjectIdTypeAdapter
import cn.twin_kles.plldb.util.JSONUtil
import com.google.gson.*
import kotlin.Throws
import java.lang.reflect.Type

object JSONUtil {

    private val GSON = GsonBuilder()
            .disableHtmlEscaping()
            .serializeNulls()
            .setDateFormat("yyyy-MM-dd HH:mm:ss")
            .registerTypeAdapter(ObjectId::class.java, ObjectIdTypeAdapter())
            .create()

    fun toTree(o: Any?): JsonElement {
        return GSON.toJsonTree(o)
    }

    @JvmStatic
	fun <T> fromTree(e: JsonElement, type: Type?): T {
        return GSON.fromJson(e, type)
    }

    @JvmStatic
	fun <T> fromString(s: String, type: Type?): T {
        return GSON.fromJson(s, type)
    }

    private class ObjectIdTypeAdapter : JsonSerializer<ObjectId>, JsonDeserializer<ObjectId> {
        override fun serialize(objectId: ObjectId, type: Type, jsonSerializationContext: JsonSerializationContext): JsonElement {
            return JsonPrimitive(objectId.toHexString())
        }

        @Throws(JsonParseException::class)
        override fun deserialize(jsonElement: JsonElement, type: Type, jsonDeserializationContext: JsonDeserializationContext): ObjectId {
            return ObjectId(jsonElement.asString)
        }
    }
}